pkg/autoscaler/autoscaler/tidb_autoscaler.go

```diff
@@ -80,3 +87,25 @@ func updateTcTiDBAnnIfScale(tac *v1alpha1.TidbClusterAutoScaler) {
 	tac.Annotations[label.AnnTiDBLastAutoScalingTimestamp] = time.Now().String()
 	emptyAutoScalingCountAnn(tac, v1alpha1.TiDBMemberType)
 }
+
+func calculateTidbMetrics(tac *v1alpha1.TidbClusterAutoScaler, sts *appsv1.StatefulSet, client promClient.Client, instances []string) (int32, error) {
+	metric := calculate.FilterMetrics(tac.Spec.TiDB.Metrics)
+	mType, err := calculate.GenMetricType(tac, metric)
+	if err != nil {
+		return 0, err
+	}
+	switch mType {
+	case calculate.MetricTypeCPU:
+		return calculate.CalculateCpuMetrics(tac, sts, client, instances, metric, calculate.TidbSumCpuMetricsPattern, *tac.Spec.TiDB.MetricsTimeDuration, v1alpha1.TiDBMemberType)
+	default:
+		return 0, fmt.Errorf(calculate.InvalidTacMetricConfigureMsg, tac.Namespace, tac.Name)
+	}
+}
+
+func filterTidbInstances(tc *v1alpha1.TidbCluster) []string {
+	var instances []string
+	for i := 0; int32(i) < tc.Spec.TiDB.Replicas; i++ {
```

Comment:
Should we get the replicas from sts considering the failover case and check if Pod is Running & Ready?

Reply:
I think currently we can leave it be, whether changing the instances filter logic depends on the future performance.